### PR TITLE
[Qwix] Quantize megablox gmm kernel

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

Quantize the megablox gmm kernel with Qwix and added unit tests

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

unit tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/kernels/megablox/gmm.py [hunk 1]
```diff
@@ -574,20 +601,18 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
       cost_estimate=cost_estimate,
   )
 
-  lhs_contracting_axis, rhs_contracting_axis = dot_general_dims[0]
+  _, rhs_contracting_axis = dot_general_dims[0]
   # Since block_spec.block_shape of rhs is None, the first axis is reduced
   # inside kernel, e.g., if block_shape is (None, tn, tk) then a tensor of
   # shape (tn, tk) will be feteched inside kernel instead of (1, tn, tk).
   # Therefore, we need to add one to rhs_contracting_axis.
   rhs_contracting_axis = map(lambda x: x + 1, rhs_contracting_axis)
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py:603]: This line is for AQT and is no longer needed.

File: MaxText/kernels/megablox/gmm.py [hunk 2]
```diff
@@ -574,20 +601,18 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
...
 
   if lhs_quantize_dtype is not None:
-    lhs_quantize_bits = 4 if lhs_quantize_dtype == jnp.int4 else 8
-    lhs = aqt_pl.quant(lhs, lhs_quantize_bits, lhs_contracting_axis)
+    lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype)
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]: To use the same channelwise quantization as AQT, we should write

```
lhs = qpl.quantize(lhs, lhs_quantize_dtype, channelwise_axes=[0])
rhs = qpl.quantize(rhs, rhs_quantize_dtype, channelwise_axes=[0, 1 if transpose_rhs else 2])
```

File: MaxText/kernels/megablox/gmm.py [hunk 3]
```diff
@@ -28,11 +28,38 @@
...
+
+def load_qarray(qa: QArray) -> QArray:
+  """
+  Materialise a `QArray` whose internal buffers are still
+  `jax.experimental.pallas.MemoryRef`s into a `QArray`
+  whose `qvalue`, `scale`, and `zero_point` fields are
+  regular `jax.Array`s.
+  """
+
+  return dataclasses.replace(
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]: This function can be written as `return jax.tree.map(lambda x: x[...], qa)`

Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py]: updated

File: MaxText/kernels/megablox/gmm.py (hunk 2 above)
Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py]: done

File: MaxText/kernels/megablox/gmm.py (hunk 1 above)
Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py:603]: removed

File: MaxText/kernels/megablox/gmm.py [hunk 4]
```diff
@@ -574,20 +585,15 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
...
-  if not isinstance(rhs, QTensor) and rhs_quantize_dtype is not None:
-    rhs_quantize_bits = 4 if rhs_quantize_dtype == jnp.int4 else 8
-    rhs = aqt_pl.quant(rhs, rhs_quantize_bits, list(rhs_contracting_axis))
+    lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype, channelwise_axes=[0])
+
+  if rhs_quantize_dtype is not None:
+    # Since block_spec.block_shape of rhs is None, the first axis is reduced
+    # inside kernel, e.g., if block_shape is (None, tn, tk) then a tensor of
+    # shape (tn, tk) will be feteched inside kernel instead of (1, tn, tk).
+    # Therefore, we need to add one to rhs_contracting_axis.
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]: The comment here is largely irrelevant. If you want some comments,

```
# Use per-channel scales for non-contracting axes, i.e., num_groups, m, n but not k.
```

File: MaxText/kernels/megablox/gmm.py [hunk 5]
```diff
@@ -400,8 +411,8 @@ def gmm(
   def kernel(
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py:407]: The "we need to know contracting axis when we quantized lhs and rhs" section above this line is also irrelevant any more.

**Review by liudangyi — APPROVED** [review]

File: MaxText/kernels/megablox/gmm.py [hunk 6]
```diff
@@ -551,7 +563,7 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
   flops = 2 * m * k * n
   cost_estimate = pl.CostEstimate(flops=flops, bytes_accessed=bytes_accessed, transcendentals=0)
   if lhs_quantize_dtype is not None or rhs_quantize_dtype is not None:
-    pallas_call_fn = aqt_pl.pallas_call
+    pallas_call_fn = qpl.pallas_call
```
Comment (RissyRan) [inline MaxText/kernels/megablox/gmm.py:573]: I heard from @KissShane that Qwix hasn't supported the tgmm quantization yet. Not sure if we'd like to hold this PR or support both till then. Otherwise, the quantization of DeepSeek will have lower perf. cc @suexu1025

File: MaxText/tests/quantizations_test.py [hunk 7]
```diff
@@ -339,5 +339,55 @@ def test_fp8_full_quantization(self):
...
+  start = 0
+  for gid, size in enumerate(group_sizes):
+    end = start + int(size)
+    ref_blocks.append(lhs[start:end] @ rhs[gid])
+    start = end
+  ref_out = jnp.concatenate(ref_blocks, axis=0)
+
+  # ---- checks ---------------------------------------------------------------
+  assert out.shape == (m, n)
+  assert jnp.allclose(out, ref_out, rtol=1e-1, atol=1e-1), "GMM output mismatch"
```
Comment (RissyRan) [inline MaxText/tests/quantizations_test.py]: It seems to me this test is checking the gmm logic, but not the quantization result vs. non-quntization (bf16)?

File: MaxText/kernels/megablox/gmm.py (hunk 6 above)
Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py:573]: `tgmm` wasn't quantized with AQT on main branch. @KissShane has it in her branch. This PR compares the perf of AQT and Qwix config and proves performance parity. I have added the results in the description. I'm planning to work on tgmm next in a separate PR.

File: MaxText/tests/quantizations_test.py (hunk 7 above)
Comment (khatwanimohit) [reply MaxText/tests/quantizations_test.py]: updated the test

**Comment by BirdsOfAFthr** [comment]
LGTM

File: MaxText/kernels/megablox/gmm.py (hunk 6 above)
Comment (RissyRan) [reply MaxText/kernels/megablox/gmm.py:573]: I see. It makes sense then!

File: MaxText/layers/moe.py [hunk 8]
```diff
@@ -639,7 +639,13 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
...
         lhs_quantize_dtype = quant_dg.fwd.dg_quantizer.lhs.numerics.get_dtype()
         rhs_quantize_dtype = quant_dg.fwd.dg_quantizer.rhs.numerics.get_dtype()
-
+      if self.config.use_qwix_quantization:
+        if self.config.quantization == "int8":
+          lhs_quantize_dtype = jnp.int8
+          rhs_quantize_dtype = jnp.int8
+        elif self.config.quantization == "fp8":
+          lhs_quantize_dtype = jnp.float8_e4m3fn
+          rhs_quantize_dtype = jnp.float8_e4m3fn
```
Comment (RissyRan) [inline MaxText/layers/moe.py]: Shall we have assertion for this `quantization` in pyconfig.py? otherwise, it could be silently out.

File: MaxText/tests/quantizations_test.py [hunk 9]
```diff
@@ -339,5 +339,53 @@ def test_fp8_full_quantization(self):
...
+      lhs,
+      rhs,
+      group_sizes,
+      tiling=tiling,  # small tiles so the shapes above work
+      interpret=True,  # avoids device-specific compilation in CI
+      lhs_quantize_dtype=dtype,
+      rhs_quantize_dtype=dtype,
+  ).block_until_ready()
+
+  assert jnp.allclose(base_out, quant_out, rtol=1e-1, atol=1e-1), "GMM output mismatch"
```
Comment (RissyRan) [inline MaxText/tests/quantizations_test.py]: Thank you for the test! @liudangyi  Wondering if this is expected tolerance? we usually see 1e-2 or 1e-3 relative and absolute tol, but not sure for quantization compared to float32. I guess it's reasonable, but would like to quick check.

**Review by RissyRan — COMMENTED** [review]
LGTM! Minor comments. Thank you!

File: MaxText/tests/quantizations_test.py (hunk 9 above)
Comment (liudangyi) [reply MaxText/tests/quantizations_test.py]: This asserts on the maximum error. I would recommend switching to assert mae or mse and we will see a more reasonable number.

File: MaxText/layers/moe.py (hunk 8 above)
Comment (khatwanimohit) [reply MaxText/layers/moe.py]: Added assert for quantization methods in pyconfig

File: MaxText/tests/quantizations_test.py (hunk 9 above)
Comment (khatwanimohit) [reply MaxText/tests/quantizations_test.py]: asserting on MAE now

**Review by jshin1394 — APPROVED** [review]

File: MaxText/kernels/megablox/gmm.py [hunk 10]
```diff
@@ -582,12 +601,21 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
...
   if lhs_quantize_dtype is not None:
-    lhs_quantize_bits = 4 if lhs_quantize_dtype == jnp.int4 else 8
-    lhs = aqt_pl.quant(lhs, lhs_quantize_bits, lhs_contracting_axis)
-
-  if not isinstance(rhs, QTensor) and rhs_quantize_dtype is not None:
-    rhs_quantize_bits = 4 if rhs_quantize_dtype == jnp.int4 else 8
-    rhs = aqt_pl.quant(rhs, rhs_quantize_bits, list(rhs_contracting_axis))
+    if use_qwix_quantization:
+      lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype, channelwise_axes=[0])
+      lhs = dataclasses.replace(lhs, scale_dtype=jnp.float32)
```
Comment (liudangyi) [inline MaxText/kernels/megablox/gmm.py]:
```
      lhs = qpl.quantize(lhs, qtype=lhs_quantize_dtype, channelwise_axes=[0], scale_dtype=jnp.float32)
```

**Review by liudangyi — APPROVED** [review]

File: MaxText/kernels/megablox/gmm.py (hunk 10 above)
Comment (khatwanimohit) [reply MaxText/kernels/megablox/gmm.py]: Done

File: MaxText/pyconfig.py [hunk 11]
```diff
@@ -241,6 +241,19 @@ def validate_constant_bound(keys):
   ), "Please specify competete constant bound or none"
 
 
+def validate_quantization_methods(keys):
```
Comment (RissyRan) [inline MaxText/pyconfig.py:244]: Thanks! Shall we also update [here](https://github.com/AI-Hypercomputer/maxtext/blob/c5bbad832d91eb0174406e976967b9a87134b3f8/MaxText/configs/base.yml#L93), I think `fp8_full` is missing.

Comment (khatwanimohit) [reply MaxText/pyconfig.py:244]: done

**Comment by SamuelMarks** [comment]
FYI: @khatwanimohit your branch will never be reached [`RoutedMoU::sparse_matmul::gmm`]:
```py
        quantization_rule = qpl.get_current_rule("dot_general")
        if quantization_rule is not None:
          lhs_quantize_dtype = quantization_rule.act_qtype
          rhs_quantize_dtype = quantization_rule.weight_qtype
```

`quantization_rule` is always `None`. Reported bug @ google/qwix/issues/79
